src/v/cluster/tx_gateway_frontend.cc

```diff
@@ -46,6 +70,892 @@ tx_gateway_frontend::tx_gateway_frontend(
+          } else {
+              (void)self.abort_tm_tx(stm, tx, timeout, outcome);
+          }
+
+          return outcome->get_future().then([](tx_errc ec) {
+              if (ec == tx_errc::none) {
+                  return kafka::end_txn_response_data{
+                    .error_code = kafka::error_code::none};
+              }
+
+              return kafka::end_txn_response_data{
+                .error_code = kafka::error_code::unknown_server_error};
+          });
+      });
+}
+
+ss::future<checked<cluster::tm_transaction, tx_errc>>
+tx_gateway_frontend::abort_tm_tx(
+  ss::shared_ptr<cluster::tm_stm>& stm,
+  cluster::tm_transaction tx,
+  model::timeout_clock::duration timeout,
+  ss::lw_shared_ptr<ss::promise<tx_errc>> outcome) {
+    return stm->get_end_lock(tx.id)->with([this, &stm, timeout, tx, outcome]() {
+        return do_abort_tm_tx(stm, tx, timeout, outcome);
+    });
+}
+
+ss::future<checked<cluster::tm_transaction, tx_errc>>
+tx_gateway_frontend::do_abort_tm_tx(
+  ss::shared_ptr<cluster::tm_stm>& stm,
+  cluster::tm_transaction tx,
+  model::timeout_clock::duration timeout,
+  ss::lw_shared_ptr<ss::promise<tx_errc>> outcome) {
+    auto changed_tx = co_await stm->try_change_status(
+      tx.id, tx.etag, cluster::tm_transaction::tx_status::aborting);
+    if (!changed_tx.has_value()) {
+        outcome->set_value(tx_errc::timeout);
+        co_return checked<cluster::tm_transaction, tx_errc>(tx_errc::timeout);
+    }
+    outcome->set_value(tx_errc::none);
+
+    tx = changed_tx.value();
+    std::vector<ss::future<abort_tx_reply>> pfs;
+    for (auto rm : tx.partitions) {
+        pfs.push_back(
+          _rm_partition_frontend.local().abort_tx(rm.ntp, tx.pid, timeout));
+    }
+    std::vector<ss::future<abort_group_tx_reply>> gfs;
+    for (auto group : tx.groups) {
+        gfs.push_back(_rm_group_frontend.local().abort_group_tx(
+          group.group_id, tx.pid, timeout));
+    }
+    auto prs = co_await when_all_succeed(pfs.begin(), pfs.end());
+    auto grs = co_await when_all_succeed(gfs.begin(), gfs.end());
+    bool ok = true;
+    for (auto r : prs) {
+        ok = ok && (r.ec == tx_errc::none);
+    }
+    for (auto r : grs) {
+        ok = ok && (r.ec == tx_errc::none);
+    }
+    if (!ok) {
+        co_return checked<cluster::tm_transaction, tx_errc>(tx_errc::timeout);
+    }
+    changed_tx = stm->mark_tx_finished(tx.id, tx.etag);
+    if (!changed_tx.has_value()) {
+        co_return checked<cluster::tm_transaction, tx_errc>(tx_errc::timeout);
+    }
+    tx = changed_tx.value();
+    co_return checked<cluster::tm_transaction, tx_errc>(tx);
+}
+
+ss::future<checked<cluster::tm_transaction, tx_errc>>
+tx_gateway_frontend::commit_tm_tx(
+  ss::shared_ptr<cluster::tm_stm>& stm,
+  cluster::tm_transaction tx,
+  model::timeout_clock::duration timeout,
+  ss::lw_shared_ptr<ss::promise<tx_errc>> outcome) {
+    return stm->get_end_lock(tx.id)->with([this, &stm, timeout, tx, outcome]() {
```

Comment:
why capture a shared pointer by reference in the lambda? take stm by value as a parameter then move it into the capture.

Reply:
I don't remember, despite being the last PR I wrote transactions using top down approach and  `tx_gateway_frontend` was among the first changes, fixing it